pkg/services/ngalert/state/state.go

```diff
@@ -195,11 +204,11 @@ func resultNormal(state *State, _ *models.AlertRule, result eval.Result, logger
 }
 
 func resultAlerting(state *State, rule *models.AlertRule, result eval.Result, logger log.Logger) {
-	switch state.State {
-	case eval.Alerting:
+	switch {
+	case state.State == eval.Alerting:
 		logger.Debug("Keeping state", "state", state.State)
 		state.Maintain(rule.IntervalSeconds, result.EvaluatedAt)
-	case eval.Pending:
+	case state.State == eval.Pending && state.Error == nil:
```

Comment:
Imagine that the current state of an alert is `Pending (error)` (meaning there have been errors and the `for_error` has not been reached) [current state is `state.State == eval.Pending` and `state.Error != nil`]

If for some reason the datasource comes to life and it is immediately firing we do NOT want the flow to enter this case statement. We want to skip it and go to the default branching instead as it is the first time it is firing. The next state will be `state.State == eval.Pending` and `state.Error == nil`

Reply:
I think this change has a bug where an alert rule that is set to Alerting on error will enter Pending (because of the For duration), and each evaluation the For duration will be reset. This means that the Pending state will never transition to Alerting, even after For has elapsed. I was able to replicate this behaviour in Grafana for this commit.

Reply:
It does not apply anymore. This behavior has changed.